### PR TITLE
[ios][local-authentication] Migrate module to Sweet API

### PR DESCRIPTION
**Pull request by fobos531**
# Why

Continuation of migration of expo modules to the new API

# How

- routine work (update podspec, module config etc.)
- 1:1 migration from ObjC to Swift

# Test Plan

Tested NCL examples on iPhone 14 Pro & iPhone SE 3rd gen (both simulators). Seems to work fine, with one remark:
- on iPhone 14 Pro authenticating without device fallback and while not enrolled returns a "usage_description" error even though `NSFaceIDUsageDescription` is defined in `Info.plist`. Is this the intended behavior, given this is a simulator?

With that said, I think it would be beneficial if you could test this on other devices (physical & simulator) to verify the correctness of the implementation.

I'd greatly appreciate any comments on the general code quality! A few comments that I have:
- I'm currently dispatching every function on the main queue. I know there's been some discussing about queues/functions in https://github.com/expo/expo/pull/19898, but I'm not sure I entirely understand the purpose of queues, so I might have used it needlessly - maybe this is something that could be documented in the module guide? i.e. the purpose of queues, when a given function should be dispatched on a given queue (and the reasoning behind it) etc. I think it would help avoid the potential issues people could run into when developing native modules. The `AsyncFunction` docs already do touch on this matter, but additional clarifications are definitely welcome.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [X] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-local-authentication/ios/LocalAuthenticationModule.swift
Comment (fobos531) [reply packages/expo-local-authentication/ios/LocalAuthenticationModule.swift:106]: Seems like the recent comment got lost, so reiterating here: would appreciate pointers on how to improve the safety of this piece of code

**Review by tsapeta — COMMENTED** [review]
> on iPhone 14 Pro authenticating without device fallback and while not enrolled returns a "usage_description" error even though NSFaceIDUsageDescription is defined in Info.plist. Is this the intended behavior, given this is a simulator?

I think that's not correct behavior, but looks like your implementation is the same as before, so I'll just take a look after merging this.

> I'm not sure I entirely understand the purpose of queues, so I might have used it needlessly - maybe this is something that could be documented in the module guide? i.e. the purpose of queues, when a given function should be dispatched on a given queue (and the reasoning behind it)

The `.main` queue refers to [DispatchQueue.main](https://developer.apple.com/documentation/dispatch/dispatchqueue/1781006-main). It's the queue that runs on the main (UI) thread and where all the UI work is done. Basically when you're doing something UI-related (updating views or view controllers) then you must dispatch it on that queue, otherwise things may crash or have unexpected behavior. Maybe this article will help: https://www.donnywals.com/appropriately-using-dispatchqueue-main
Good point that we should note that in the docs, thanks!
In case of local authentication, it seems that we don't need to run those functions on the main thread and in the old implementation everything is run on the background thread — it doesn't override `- (dispatch_queue_t)methodQueue`. That being said, I think you can remove `.runOnQueue(.main)` from all these functions 😉 

Could you also reindent all Swift files? 😅 We prefer using 2 spaces 🙂

File: packages/expo-local-authentication/ios/LocalAuthenticationModule.swift
Comment (fobos531) [reply packages/expo-local-authentication/ios/LocalAuthenticationModule.swift]: Makes sense! Nice, optional binding is indeed a pretty elegant way to solve this 😉

**Comment by fobos531** [comment]
Hey @tsapeta , thanks for the awesome and thorough explanation, much appreciated!

I've applied the code changes you suggested 🚀

File: packages/expo-local-authentication/ios/LocalAuthenticationModule.swift
Comment (tsapeta) [inline packages/expo-local-authentication/ios/LocalAuthenticationModule.swift:17]: Oops, looks like still only Name and the first AsyncFunction are indented correctly 😅

Comment (fobos531) [reply packages/expo-local-authentication/ios/LocalAuthenticationModule.swift]: woops, should be good now :D

**Review by tsapeta — APPROVED** [review]
It's perfect now 🚀 Thank you so much for your help 🙏 
I'll wait for the CI and then merge 😉

**Review by tsapeta — APPROVED** [review]

**Comment by lexer** [comment]
@fobos531 I believe this PR introduced a bug and it's not longer working on some devices.

**Comment by lexer** [comment]
Tried to report it here. https://github.com/expo/expo/issues/21494

**Comment by lexer** [comment]
Try to call:

```
    return await LocalAuthentication.authenticateAsync({
      promptMessage: 'test',
      disableDeviceFallback: true,
      requireConfirmation: true,
      fallbackLabel: '',
      cancelLabel: '',
    });
```    
iOS version: 16.2
Model: iPhone 12 mini

**Pull request merged**
